lightbox: Added user's avatar in the lightbox app bar

shivanshsharma13 opened the pull request:
In the lightbox, the RN app shows the message author's avatar along with their name in the app bar. This change implements the same behavior in the Flutter app, positioning the avatar beside the name/date text in the "start" direction (left in LTR layouts, right in RTL layouts).

The changes:
- Add Avatar widget in the lightbox app bar next to the author's name and date
- Ensure the avatar is positioned correctly based on the text direction (LTR/RTL)
- Maintain consistent sizing and spacing (35px size, 8px right padding)
- Handle overflow gracefully for both avatars and names

Screenshot-

<img src="https://github.com/user-attachments/assets/b756b95f-53b4-4cb0-b3c3-2cef1b995563" width="400"/>

Fixes #41

alya commented:
Thanks @shivanshsharma13 ! Please add screenshots showing your changes to the PR description.

shivanshsharma13 commented:
> Thanks @shivanshsharma13 ! Please add screenshots showing your changes to the PR description.

HI @alya Done!

gnprice commented:
Thanks for the PR. Before we can review this, it will need (a) tests and (b) to follow our Git style. See our README at: https://github.com/zulip/zulip-flutter#submitting-a-pull-request

gnprice commented:
Oh also:
* The screenshot doesn't show any recipient headers. The thing at the top of the screen (with a back button and the page's title) is the app bar.

  A Zulip "recipient header" is something shown inside the message list, potentially many times, as we show a new recipient header at each point where the recipient changes from one message to the next. If you look at a combined feed you'll see many examples. See also our code — we use the term "recipient header" there.
* The linked issue is about the lightbox, which this doesn't touch. So this shouldn't be marked as fixing that issue.

shivanshsharma13 commented:
Hi, sorry was a bit confused. Now I have implemented the correct code. It's ready for review. Adding test for this as well.

chrisbobbe commented:
> Thanks for the PR. Before we can review this, it will need (a) tests and (b) to follow our Git style. See our README at: https://github.com/zulip/zulip-flutter#submitting-a-pull-request

shivanshsharma13 commented:
> > Thanks for the PR. Before we can review this, it will need (a) tests and (b) to follow our Git style. See our README at: https://github.com/zulip/zulip-flutter#submitting-a-pull-request

Done the changes!
- [X] Wrapped commit message in 68 chars a line
- [X] Added the `Fixes:` format like other commits
- [X] Added the test case for the change.

chrisbobbe commented:
Thanks for those screenshots, although I notice they're a little pixelated.

> For the Avatar I chose to use 40 as it feels more fitting and consistent with the RN app.

As I said, the RN app uses 36 and 36 / 8. The lightbox page will likely change when our designer makes a design for it. Temporarily, I think taking these numbers from the RN app is fine.

There's no point using the `n / 8` form if `n` doesn't equal the size, whether that's 36 or 40. I like 36, and it's easy to explain because it just follow the number in zulip-mobile. Greg may have a different preference.

Please choose either:

- `36` and `36 / 8` or
- `40` and `40 / 8`

Please also update the commit with the other things I requested: https://github.com/zulip/zulip-flutter/pull/1092#issuecomment-2537237037

shivanshsharma13 commented:
Hi, Thanks for the detailed explanation. Added the 36 size and 36/8 as per suggestion. Also changed the indentation. It's ready for review.🙂

chrisbobbe commented:
CI has caught a failing test; please fix that.

shivanshsharma13 commented:
Sorry, I just missed updating the test file earlier. It's done now. Thanks!

PIG208 commented on lib/widgets/lightbox.dart:
```diff
@@ -171,22 +171,28 @@ class _LightboxPageLayoutState extends State<_LightboxPageLayout> {
+                    text: timestampText,
+                    // Make smaller, like a subtitle
+                    style: themeData.textTheme.titleSmall!.copyWith(color: appBarForegroundColor)),
+                ])
+              ),
+            ),
+          ],
+        ),
```
nit: Looks like this change expands the parentheses into separate lines. We should fix this.
```suggestion
                ]))),
          ]),
```

PIG208 commented on lib/widgets/lightbox.dart:
```diff
@@ -171,22 +171,28 @@ class _LightboxPageLayoutState extends State<_LightboxPageLayout> {
+          ],
+        ),
+        bottom: widget.buildAppBarBottom(context)
+      );
```
nit: same issue here

PIG208 replied:
bump

PIG208 commented on test/widgets/lightbox_test.dart:
```diff
@@ -236,14 +236,19 @@ void main() {
       await setupPage(tester, message: message, thumbnailUrl: null);
 
-      // We're looking for a RichText, in the app bar, with both the
-      // sender's name and the timestamp.
```
nit: the comment was specifically for explaining what `RichText` was used for. Let's leave it as-is

PIG208 replied:
The comment belongs to the part started with `final labelTextWidget`, so it should come before that and replace `// Check sender name and timestamp`.

PIG208 commented on test/widgets/lightbox_test.dart:
```diff
@@ -236,14 +236,19 @@ void main() {
+      // Check Avatar properties
+      final avatar = tester.widget<Avatar>(find.byType(Avatar));
+      check(avatar.size).equals(36);
+      check(avatar.borderRadius).equals(36 / 8);
```
The design properties like `size` and `borderRadius` are usually verified during reviews. We typically don't need test for these because it only captures a limited scope of bugs and becomes another place to update if the design spec changes.

shivanshsharma13 replied:
Thank you for the review! Should I remove the test, or would you suggest modifying it in some way? Looking forward to your guidance!

PIG208 reviewed (commented):
Thanks for the update @shivanshsharma13! Left some comments. Also a commit message nit:

```
 Lightbox: Added user's avatar in the lightbox app bar 
```
to
```
 lightbox: Added user's avatar in the lightbox app bar 
```

For finding the right prefix, `git log --oneline -- lib/widgets/lightbox.dart` is handy.

shivanshsharma13 commented:
Thanks for the  detailed review. I acknowledged the suggested changes. Please let me know if anything needs to be changed!🙂

PIG208 reviewed (commented):
Thanks for the update! Left some comments.

PIG208 commented on lib/widgets/lightbox.dart:
```diff
@@ -171,22 +171,24 @@ class _LightboxPageLayoutState extends State<_LightboxPageLayout> {
+                    style: themeData.textTheme.titleSmall!.copyWith(color: appBarForegroundColor)),
+                ]))),
+          ]),
+          bottom: widget.buildAppBarBottom(context));
```
nit: Hmm, didn't notice this until now — `bottom` is a parameter for `AppBar`, so it should be at the same indent level as other parameters like `title`

PIG208 reviewed (commented):
One more commit message nit:
```
lightbox: Added user's avatar in the lightbox app bar
```

See the commit style guide linked in our README: https://github.com/zulip/zulip-flutter?tab=readme-ov-file#submitting-a-pull-request

Commit summaries should be started with an imperative verb:

```
lightbox: Add user's avatar on the lightbox app bar
```

gnprice commented on lib/widgets/lightbox.dart:
```diff
@@ -171,21 +171,23 @@ class _LightboxPageLayoutState extends State<_LightboxPageLayout> {
+                    // Make smaller, like a subtitle
+                    style: themeData.textTheme.titleSmall!.copyWith(color: appBarForegroundColor)),
+                ]))),
+          ]),
```
nit:
```suggestion
        ]),
```

The close-bracket gets aligned with the start of the line the open-bracket is on. Same thing as for braces `{}` and parens `()`.

gnprice commented on test/widgets/lightbox_test.dart:
```diff
@@ -236,12 +236,16 @@ void main() {
       final message = eg.streamMessage(sender: eg.otherUser, timestamp: timestamp);
       await setupPage(tester, message: message, thumbnailUrl: null);
 
+      // Check Avatar properties
+      final avatar = tester.widget<Avatar>(find.byType(Avatar));
+      check(avatar.userId).equals(message.senderId);
```
```suggestion
      final avatar = tester.widget<Avatar>(find.byType(Avatar));
      check(avatar.userId).equals(message.senderId);
```

This comment doesn't say anything the code just under it doesn't already say just as clearly.

It's also wrong — this code checks only one property.

I believe this illustrates the classic way that comments tend to become wrong over time, which is that you edited the code (https://github.com/zulip/zulip-flutter/pull/1092#discussion_r1889357918) and didn't update the comment. If that can happen when you yourself are editing the code, within the same PR where you wrote the comment in the first place, you can imagine it happens all the time when other people are updating the code months and years later.

The fact that comments routinely become wrong is a major reason to leave out comments except those that pull their weight by really adding something that's not easy to see by just looking at the code.

shivanshsharma13 replied:
That makes sense. I apologize for the oversight and will ensure to be more mindful of this in the future.

gnprice reviewed (commented):
Thanks @shivanshsharma13, and thanks @chrisbobbe and @PIG208 for the previous reviews! Comments below.

shivanshsharma13 commented on lib/widgets/lightbox.dart:
```diff
@@ -171,21 +171,23 @@ class _LightboxPageLayoutState extends State<_LightboxPageLayout> {
-          ])),
+        title: Row(children: [
+          Avatar(size: 36, borderRadius: 36 / 8, userId: widget.message.senderId),
+          const SizedBox(width: 8),
```
Yeah I took the reference from Profile page (_UserWidget in Profile.dart) also (_MentionAutocompleteItem in autocomplete.dart) 

For the suggested changes, please verify this..

| spacing | Image |
|-------------------------|-------|
| 12px                 | <img src="https://github.com/user-attachments/assets/fad9f47b-507c-454e-b181-c0f6e3baf661" alt="zulip_mobile" width="400" style="border-radius: 8px;"> |
| 16px  | <img src="https://github.com/user-attachments/assets/57a6fef9-f187-42d1-85f9-366b3a505d22" alt="flutter_image" width="400" style="border-radius: 8px;"> |

gnprice replied:
> Yeah I took the reference from Profile page (_UserWidget in Profile.dart) also (_MentionAutocompleteItem in autocomplete.dart)

Ah, very helpful, thanks.

And in those screenshots I see 16px is definitely too wide, and 12px already looks a little loose. Let's go with 8px, then. At some point our designer will get to this screen (he has plenty of others in his queue first), and then we'll probably change a bunch of aspects including these details, but that should be fine until then.

shivanshsharma13 commented on lib/widgets/lightbox.dart:
```diff
@@ -171,21 +171,23 @@ class _LightboxPageLayoutState extends State<_LightboxPageLayout> {
-              style: themeData.textTheme.titleSmall!.copyWith(color: appBarForegroundColor)),
-          ])),
+        title: Row(children: [
+          Avatar(size: 36, borderRadius: 36 / 8, userId: widget.message.senderId),
+          const SizedBox(width: 8),
+          Expanded(
+            child: RichText(
```
Yeah, This is the updated screenshot...

![latest_changes](https://github.com/user-attachments/assets/8d39415e-349a-4a81-9635-51939e348508)

shivanshsharma13 commented:
Hi, thank you for the detailed review. I have acknowledged all the suggested changes and attached a screenshot reflecting the recommended updates. Please let me know if there are any additional modifications required.

gnprice commented on lib/widgets/lightbox.dart:
```diff
@@ -171,21 +171,22 @@ class _LightboxPageLayoutState extends State<_LightboxPageLayout> {
+          const SizedBox(width: 8),
+          Expanded(
+            child: RichText(
+              text: TextSpan(children: [
+                  TextSpan(
+                    text: '${widget.message.senderFullName}\n',
+                    // Restate default
```
nit: avoid making changes that aren't part of what you're trying to do — this code is just getting indented one more level, so doesn't need other changes:
```suggestion
              text: TextSpan(children: [
                TextSpan(
                  text: '${widget.message.senderFullName}\n',

                  // Restate default
```

gnprice reviewed (commented):
Thanks for the revision, and the screenshots! All looks good now except one nit below. I'll fix it and merge.
